src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.DocumentChangeTracker.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    internal partial class RequestExecutionQueue
+    {
+        /// <summary>
+        /// Keeps track of changes to documents that are opened in the LSP client. Calls MUST not overlap, so this
+        /// should be called from a mutating request handler. See <see cref="RequestExecutionQueue"/> for more details.
+        /// </summary>
+        internal class DocumentChangeTracker : IWorkspaceService
+        {
+            private readonly Dictionary<(Workspace Workspace, DocumentId Id), SourceText> _trackedDocuments = new();
```

Comment:
> Do we have workspace in the key to avoid issues with the same documentId in different workspaces?

I had the same question on the [pull diagnostics PR](https://github.com/dotnet/roslyn/pull/48328/files#diff-746af321ae8dde2dee926532a1b6ff5e20168af0dcfd2a50b86bb97d4cba28c6R36) and the answer was yes.

Reply:
As mentioned during the meeting today, this probably shouldn't be including Workspace in the key at all, especially if a document needs to transition between workspaces but maintain content.

To answer the question more generally: good question! So generally even if a Workspace is disposed, existing snapshots are still valid, and calling .CurrentSolution still works too. So if we were holding onto a Workspace we wouldn't end up with extreme badness like crashes....but it'd still be a memory leak. We don't have a Disposed event so we'd either have to go sort one out or you'd have to refactor some stuff to be either a ConditionalWeakTable or use a workspace service somehow to hold onto the workspace-specific state.

Reply:
Yeah I'm going to have to rethink this somewhat. Document IDs aren't unique to a workspace, so can't use them, but also it's easier to have them so I can easily find the document in the solution later, to update it's contents.

Uri/filepath is the best thing to use for caching, but to go from that to a solution we have to ask each workspace, which we can only do after getting the solution for the request, which we get by asking each workspace!

Reply:
Okay, worked this out, I was overthinking. I already know what solution we care about, so don't need to ask each workspace.